[WIP] Update buildroot to 2025.11-rc2 and allow building for gcc-15 hosts

Knogle opened the pull request:
Ahoy, maybe consider this as a WIP PR?
I've updated the buildroot version to 2025.11-rc2 in order to properly build with gcc-15. Also another patch for kbd is necessary in order to build without errors.

Also we had to increase the EFI partition size, but i don't know if we can get a way around it. Maybe we have to create new defconfigs.
We also have to get rid of legacy parts in the defconfig as they are no longer valid.

desertwitch committed:
feat(iso9660): complete refactor of filesystem

The entire iso9660 filesystem folder was restructured to replace partially unreadable legacy code with modern, readable and modular structures. Hardcoded paths were replaced with variables, as much as possible, to make the code work with different setups. A mode "both" was added, producing Debian-style hybrid ISOs for both BIOS and EFI usage, while still allowing for flashing to USB and/or burning to CD/DVD-ROM. The configuration GUI (via "make menuconfig") was also refactored and adapted to the new changes.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(iso9660): complete refactor of filesystem

The entire iso9660 filesystem folder was restructured to replace partially unreadable legacy code with modern, readable and modular structures. Hardcoded paths were replaced with variables, as much as possible, to make the code work with different setups. A mode "both" was added, producing Debian-style hybrid ISOs for both BIOS and EFI usage, while still allowing for flashing to USB and/or burning to CD/DVD-ROM. The configuration GUI (via "make menuconfig") was also refactored and adapted to the new changes.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore(grub2): remove obsolete vanity patch

As the board is now using a GRUB menu, replacing these messages no longer makes sense (most users will never see them), so the additional patch is better removed.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore(grub2): remove obsolete vanity patch

As the board is now using a GRUB menu, replacing these messages no longer makes sense (most users will never see them), so the additional patch is better removed.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(shredos): refactor post-process scripts

Post-processing used a static (outdated?) GRUB image for USB image generation, but it is better to use the GRUB image produced as part of the building process, which ensures that the actual compiled GRUB version is used as bootloader in our images and not an outdated one.

Post-processing previously embedded the 64-bit EFI bootloader to 32-bit builds, as no architecture checks were in place. This was changed to check for architecture and embed the correct EFI bootloader for the respective target architecture instead.

The GRUB configuration was changed to use a menu, to allow the user to choose between standard and nomodeset parameters, while also allowing the user to edit the command line to append other kernel parameters they may need for their systems.

The vanity kernel image name was returned to bzImage to clearly mark the image for what it is, and to remain consistent with the ISO file generation where we have no direct control over the image now.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(shredos): refactor post-process scripts

Post-processing used a static (outdated?) GRUB image for USB image generation, but it is better to use the GRUB image produced as part of the building process, which ensures that the actual compiled GRUB version is used as bootloader in our images and not an outdated one.

Post-processing previously embedded the 64-bit EFI bootloader to 32-bit builds, as no architecture checks were in place. This was changed to check for architecture and embed the correct EFI bootloader for the respective target architecture instead.

The GRUB configuration was changed to use a menu, to allow the user to choose between standard and nomodeset parameters, while also allowing the user to edit the command line to append other kernel parameters they may need for their systems.

The vanity kernel image name was returned to bzImage to clearly mark the image for what it is, and to remain consistent with the ISO file generation where we have no direct control over the image now.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(configs): adapt configurations for new changes

The default configuration "shredos_defconfig" was amended to produce both the existing USB image (which allows writing PDFs to the USB) and the new hybrid ISO (which now allows BIOS+UEFI booting, as well as flashing to USB and burning to CD/DVD-ROM) in one build.

Additional configurations "shredos_img_defconfig" and "shredos_iso_defconfig" were added which only target either the existing USB image or the new hybrid ISO, for full flexibility.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(configs): adapt configurations for new changes

The default configuration "shredos_defconfig" was amended to produce both the existing USB image (which allows writing PDFs to the USB) and the new hybrid ISO (which now allows BIOS+UEFI booting, as well as flashing to USB and burning to CD/DVD-ROM) in one build.

Additional configurations "shredos_img_defconfig" and "shredos_iso_defconfig" were added which only target either the existing USB image or the new hybrid ISO, for full flexibility.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(configs): add i586-specific configurations

Additional configurations for the i586 builds were added, so that new users can easily build those targets without needing to change things around via "make menuconfig" (potentially triggering unwanted other changes, e.g. reset of GRUB modules).

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(configs): add i586-specific configurations

Additional configurations for the i586 builds were added, so that new users can easily build those targets without needing to change things around via "make menuconfig" (potentially triggering unwanted other changes, e.g. reset of GRUB modules).

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: update readme for new changes

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: update readme for new changes

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): improve xorriso argument order

Improved the iso9660 argument order to better reflect modern best practices, inspired by the Debian distribution ISOs.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): improve xorriso argument order

Improved the iso9660 argument order to better reflect modern best practices, inspired by the Debian distribution ISOs.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): default path adapted for clarity

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): default path adapted for clarity

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(configs): add missing grub modules

We now require configfile and search modules to locate the ISO9660 filesystem, add them for our targets.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(configs): add missing grub modules

We now require configfile and search modules to locate the ISO9660 filesystem, add them for our targets.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): copy actual efi config to efi structure

To keep things consistent, copy the EFI configuration also to the EFI structure on the ISO9660 filesystem.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): copy actual efi config to efi structure

To keep things consistent, copy the EFI configuration also to the EFI structure on the ISO9660 filesystem.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: clarification in readme

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: clarification in readme

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(configs,shredos): add aio iso configuration

This adds a configuration for a jack-of-all-trades ISO, which allows BIOS and UEFI booting, being written to CD/DVD-ROM and USB drives, as well as having a writeable partition when being used from an USB drive. It basically combines the functionality of all prior release versions into one single ISO (per architecture).

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(configs,shredos): add aio iso configuration

This adds a configuration for a jack-of-all-trades ISO, which allows BIOS and UEFI booting, being written to CD/DVD-ROM and USB drives, as well as having a writeable partition when being used from an USB drive. It basically combines the functionality of all prior release versions into one single ISO (per architecture).

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: update readme for new changes

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: update readme for new changes

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore(iso9660): add partition size documentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore(iso9660): add partition size documentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(configs): make default configs build USB image and AIO hybrid ISO

The all-in-one ISO works well and is the jack-of-all-trades, covering all imaginable systems and scenarios, so include that in the default build configurations.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(configs): make default configs build USB image and AIO hybrid ISO

The all-in-one ISO works well and is the jack-of-all-trades, covering all imaginable systems and scenarios, so include that in the default build configurations.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat: add basic multi-build script

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat: add basic multi-build script

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix: embed grub finder configuration

Helps GRUB find the ShredOS volume when it otherwise cannot (such as when the IMG is used with Ventoy in BIOS)

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix: embed grub finder configuration

Helps GRUB find the ShredOS volume when it otherwise cannot (such as when the IMG is used with Ventoy in BIOS)

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: update buildscript and documentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: update buildscript and documentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: add warning about customization to ISO

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: add warning about customization to ISO

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: update documentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: update documentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix: amend buildscript

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix: amend buildscript

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(buildscript): use reconfigure instead of rebuild

Reconfigure starts one stage before rebuild, just to be safe. It is the earliest step after the source download and patching.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(buildscript): use reconfigure instead of rebuild

Reconfigure starts one stage before rebuild, just to be safe. It is the earliest step after the source download and patching.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: reorganize all files into board/shredos

It is more convenient to not have to search for configuration files in different directories, and also helpful to see the standard configuration files, so move all ShredOS configurations into board/shredos.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: reorganize all files into board/shredos

It is more convenient to not have to search for configuration files in different directories, and also helpful to see the standard configuration files, so move all ShredOS configurations into board/shredos.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): account for both mode without hybrid

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): account for both mode without hybrid

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore(configs): clean up configurations

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore(configs): clean up configurations

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(configs): revert partition size and post processing for EFI

The EFI partition for hybrid ISOs does not reliably show on all Windows systems, so we cannot use our "hack" of writing the configurations/PDFs to it. Reverted instead to a sane-sized EFI partition which is used just for booting. Added a note to the README that some tools such as Rufus support an ISO-mode to add a writeable partition at burning-time.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(configs): revert partition size and post processing for EFI

The EFI partition for hybrid ISOs does not reliably show on all Windows systems, so we cannot use our "hack" of writing the configurations/PDFs to it. Reverted instead to a sane-sized EFI partition which is used just for booting. Added a note to the README that some tools such as Rufus support an ISO-mode to add a writeable partition at burning-time.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(iso9660): appendable partition image for hybrid modes

Allow to append a FAT16 partition image when building an ISO in hybrid mode.

This allows to add an extra writeable partition for when the ISO is burned to USB.

This feature is experimental and disabled by default.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(iso9660): appendable partition image for hybrid modes

Allow to append a FAT16 partition image when building an ISO in hybrid mode.

This allows to add an extra writeable partition for when the ISO is burned to USB.

This feature is experimental and disabled by default.

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(iso9660): add volume name and partition type as options

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
feat(iso9660): add volume name and partition type as options

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: adapt documentation for current state

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: adapt documentation for current state

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(configs): add missing argument

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(configs): add missing argument

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: adapt documentation for current state

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: adapt documentation for current state

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): use buildroot rm, adapt comment

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): use buildroot rm, adapt comment

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): menu ordering

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): menu ordering

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): menu indentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
fix(iso9660): menu indentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: clarification in documentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

desertwitch committed:
chore: clarification in documentation

Signed-off-by: desertwitch <24509509+desertwitch@users.noreply.github.com>

Knogle committed:
board/shredos: enable crypto options for future AF_ALG AES-CTR support

Enable additional cryptographic kernel options required for using
AES-CTR via the AF_ALG userspace interface in future features:

- CONFIG_CRYPTO_PCRYPT: parallel crypto wrapper for symmetric ciphers
- CONFIG_CRYPTO_USER_API_SKCIPHER: AF_ALG user API for skcipher
- CONFIG_CRYPTO_AES_NI_INTEL: AES-NI accelerated AES implementation on x86_64

These options prepare the kernel for an AF_ALG-based AES-CTR PRNG and
other potential AF_ALG-backed cryptographic functionality in nwipe.

bkuhls committed:
package/kbd: fix zlib build, add optional dependencies

Buildroot commit 930660890bf8b40d9bd55cd9e202e574b53d5d49 bumped kbd to
version 2.9.0 which includes optional support for various compression
libraries. This commit adds the corresponding configure options and a
fix for build errors caused by zlib.

Fixes:
https://autobuild.buildroot.net/results/8ff/8ff6c3d940b68069f748f12646f7516ec86172c1/

Signed-off-by: Bernd Kuhls <bernd@kuhls.net>

Knogle committed:
Vendor drop: Update Buildroot core to 2025.11-rc2 and re-add ShredOS board/packages

Knogle committed:
fs/iso9660: increase EFI system partition image size

Recent builds may fail with "Disk full" when generating the EFI FAT
partition image (fat.efi). The iso9660 filesystem backend creates the
EFI system partition as a 1 MiB FAT filesystem and then copies all EFI
boot files from output/images/efi-part/ into it using mcopy. On systems
with larger EFI payloads (e.g. GRUB binaries, shim, drivers or vendor
files), the 1 MiB filesystem is insufficient and the build stops with:

    mcopy: Disk full

This commit increases the FAT image size from 1 MiB to 16 MiB, providing
enough space for typical EFI bootloader contents and preventing the
filesystem generation step from failing.

No functional change to runtime behaviour; only the size of the embedded
EFI system partition image is increased.

desertwitch commented:
You shouldn't use their iso9660 version, it's still broken and hasn't been updated in years. Instead, port over the iso9660 folder we're using and you won't need to hardcore the EFI partition size because it's a configurable now (defaulting to 3M). Our defconfigs won't work with their iso9660 version either, because multiple new configurables were introduced for iso9660 by us.

Just copy over the whole iso9660 folder.

See also: https://github.com/PartialVolume/shredos.x86_64/pull/393#issuecomment-3573815361

Knogle commented:
> You shouldn't use their iso9660 version, it's still broken and hasn't been updated in years. Instead, port over the iso9660 folder we're using and you won't need to hardcore the EFI partition size because it's a configurable now (defaulting to 3M). Our defconfigs won't work with their iso9660 version either, because multiple new configurables were introduced for iso9660 by us.
> 
> Just copy over the whole iso9660 folder.
> 
> See also: [#393 (comment)](https://github.com/PartialVolume/shredos.x86_64/pull/393#issuecomment-3573815361)

Ahhh great info! Thanks! I will do so and commit again.

Knogle commented:
I hope it worked. It's like the first time i am doing `cherry-picking` using `git`.

desertwitch commented:
Seems to have worked, `git diff` shows the directory as identical to the current `master` (which we want).
I'm afraid I have no time to test this further, just wanted to inform about the `iso9660` needing porting over.

Knogle committed:
package/unzip: fix conflicting gmtime/localtime prototypes with GCC 15

GCC 15 enforces stricter prototype checks for standard library
functions. The unzip 6.0 sources declare their own prototypes for
gmtime() and localtime() in unix/unxcfg.h without parameters, which
conflicts with the libc-provided declarations that expect a `const time_t *`.

This results in the following build failures:

    error: conflicting types for 'gmtime'
    error: conflicting types for 'localtime'
    error: too many arguments to function 'localtime'
    error: too many arguments to function 'gmtime'

The custom prototypes in unxcfg.h are unnecessary on modern systems,
and removing them restores correct type checking and fixes the build
with GCC 15.

This patch deletes the outdated declarations:

    struct tm *gmtime(), *localtime();

Tested-by: Fabian Druschke fdruschke@outlook.com
Signed-off-by: Fabian Druschke fdruschke@outlook.com

PartialVolume commented:
I hate to be the bearer of bad news but I probably won't use this. It's just too large and too many changes all in one PR that are not related and makes testing a pain. When I do a buildroot  update I use some code to compare old configs with new to makesure nothing has been introduced that shouldn't be there but also nothing is missing from the 3-4 thousand programs build into ShredOS.

Knogle commented:
> I hate to be the bearer of bad news but I probably won't use this. It's just too large and too many changes all in one PR that are not related and makes testing a pain. When I do a buildroot update I use some code to compare old configs with new to makesure nothing has been introduced that shouldn't be there but also nothing is missing from the 3-4 thousand programs build into ShredOS.

Oh makes sense! No problem though, it was just more out of interest and for personal testing :) Thank you though for your reply! I'll close this one here.

Buildroot is honestly a tough one to update, and many things break as we see here as well.

PartialVolume commented:
@Knogle In regards to gcc, last night I tested the current master of nwipe against Debian Testing, which has gcc version 15.2.0 (Debian 15.2.0.9) using the default ./configure with no additional options and it built without any errors. 

I also tried
```
Martijn's preferred command for testing on Debian SID
./configure --prefix=/usr CFLAGS='-g -O2 -fstack-protector-strong -Wformat -Werror=format-security
```

No errors or warnings using the above command either.

And finally
```
./configure --prefix=/usr CFLAGS='-O0 -g -Wall -Wextra -fstack-protector-strong -Wformat -Werror=format-security' 
```
Again no errors or warnings. You submitted a gcc15 fix PR that went into v0.39, was originally submitted by @xandris but closed for some reason even though it was valid, hence why you later re-commited it.

I seemed to remember you recently saying something to do with Fedora and gcc 15? But can't find that comment. Maybe had I tested nwipe with Fedora?

Is there a build issue with nwipe and gcc 15? as I'm not seeing it on Debian Testing.

(This is probably the wrong repository for this comment, but as you mentioned gcc15 in this PR it reminded me to mention it.)

Knogle commented:
Ahoy! With gcc-15 there is no build issue anymore with nwipe. Someone had created a PR sometime ago to fix some declarations in Mersenne and Isaac I think to make it work.
What currently causes issues with gcc-15 I think is only ShredOS at the moment.

EDIT: Yep. Using the fix made by @xandris it works fine with gcc-15
